### PR TITLE
move resource builder flags to genericclioptions

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

Moves resource builder flags to genericclioptions

cc @soltysh @deads2k

### CODE REVIEW

**Comment by deads2k** [comment]
When you do the move, you need to make things optional.

File: pkg/kubectl/genericclioptions/filename_flags.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"k8s.io/kubernetes/pkg/kubectl/genericclioptions/resource"
+)
+
+type FileNameFlags struct {
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/filename_flags.go]: Indicate in doc that usage of this is discouraged.  Use the resourcebuilder flags

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 2]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
+	}
 }
 
 // ToBuilder gives you back a resource finder to visit resources that are located
-func (o *ResourceBuilderFlags) ToBuilder(restClientGetter genericclioptions.RESTClientGetter, resources []string) ResourceFinder {
+func (o *ResourceBuilderFlags) ToBuilder(restClientGetter RESTClientGetter, resources []string) ResourceFinder {
 	namespace, enforceNamespace, namespaceErr := restClientGetter.ToRawKubeConfigLoader().Namespace()
 
+	labelSelector := ""
+	if o.LabelSelector != nil {
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: No, you need to use these ifs to decide whether to call the builder methods at  all.

```go
if o.labelselector != nil{
 builder.LabelSelectorParam()
}
```

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 3]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
 
 // ResourceBuilderFlags are flags for finding resources
 type ResourceBuilderFlags struct {
-	FilenameOptions resource.FilenameOptions
+	FileNameFlags *FileNameFlags
 
-	LabelSelector     string
-	FieldSelector     string
-	AllNamespaces     bool
 	Namespace         string
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: used or dead?

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 4]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
 	ExplicitNamespace bool
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: used or dead?

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 5]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
 
-	// TODO add conditional support.  These are false for now.
-	All   bool
-	Local bool
+	LabelSelector *string
+	FieldSelector *string
+	AllNamespaces *bool
+
+	All   *bool
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: used or dead?  If dead.  Wire up in a followup

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 6]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
+	Local *bool
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: used or dead?  If dead.  Wire up in a followup

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 7]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
+			Usage:     "identifying the resource.",
+			Filenames: &filenames,
+			Recursive: bool_ptr(true),
 		},
+
+		LabelSelector: str_ptr(""),
+		FieldSelector: str_ptr(""),
+		AllNamespaces: bool_ptr(false),
+
+		All:   bool_ptr(false),
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: remove

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 8]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
+		Local: bool_ptr(false),
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: remove

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 9]
```diff
@@ -14,68 +14,95 @@ See the License for the specific language governing permissions and
...
-		FilenameOptions: resource.FilenameOptions{
-			Recursive: true,
+		FileNameFlags: &FileNameFlags{
+			Usage:     "identifying the resource.",
+			Filenames: &filenames,
+			Recursive: bool_ptr(true),
 		},
+
+		LabelSelector: str_ptr(""),
+		FieldSelector: str_ptr(""),
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/kubectl/genericclioptions/builder_flags.go]: `WithFieldSelector(string)` if we need it

File: pkg/kubectl/genericclioptions/builder_flags.go (hunk 5 above)
Comment (juanvallejo) [reply pkg/kubectl/genericclioptions/builder_flags.go]: Used in the `ToBuilder` method as an argument to the `ResourceTypeOrNameArgs` param

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 10]
```diff
@@ -112,3 +131,11 @@ func ResourceFinderForResult(result resource.Visitor) ResourceFinder {
 		return result
 	})
 }
+
+func str_ptr(val string) *string {
```
Comment (soltysh) [inline pkg/kubectl/genericclioptions/builder_flags.go]: s/str_ptr/strPtr

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 11]
```diff
@@ -112,3 +131,11 @@ func ResourceFinderForResult(result resource.Visitor) ResourceFinder {
...
+	return &val
+}
+
+func bool_ptr(val bool) *bool {
```
Comment (soltysh) [inline pkg/kubectl/genericclioptions/builder_flags.go]: s/bool_ptr/boolPtr
don't use underscore in names

**Review by soltysh — APPROVED** [review]
a few small nits

File: pkg/kubectl/genericclioptions/builder_flags.go [hunk 12]
```diff
@@ -14,68 +14,87 @@ See the License for the specific language governing permissions and
...
-	Namespace         string
-	ExplicitNamespace bool
+	LabelSelector *string
+	FieldSelector *string
+	AllNamespaces *bool
 
-	// TODO add conditional support.  These are false for now.
-	All   bool
-	Local bool
+	All *bool
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/builder_flags.go]: this doesn't have an option to be bound.

File: pkg/kubectl/genericclioptions/filename_flags.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+// These flags are composed by ResourceBuilderFlags
+// which should be used instead.
+type FileNameFlags struct {
+	Usage string
+
+	Filenames *[]string
+	Recursive *bool
+}
+
+func (o *FileNameFlags) ToOptions() resource.FilenameOptions {
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/filename_flags.go]: tolerate nil

File: pkg/kubectl/genericclioptions/filename_flags.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	if o.Filenames != nil {
+		options.Filenames = *o.Filenames
+	}
+
+	return options
+}
+
+func (o *FileNameFlags) AddFlags(flags *pflag.FlagSet) {
+	if o.Recursive != nil {
```
Comment (deads2k) [inline pkg/kubectl/genericclioptions/filename_flags.go]: tolerate nil

File: pkg/kubectl/genericclioptions/builder_flags.go (hunk 12 above)
Comment (juanvallejo) [reply pkg/kubectl/genericclioptions/builder_flags.go]: will turn into value

File: pkg/kubectl/genericclioptions/filename_flags.go (hunk 13 above)
Comment (juanvallejo) [reply pkg/kubectl/genericclioptions/filename_flags.go]: Done

**Comment by deads2k** [comment]
Needs rebase

**Comment by soltysh** [comment]
Superseded by https://github.com/kubernetes/kubernetes/pull/64390.

**Pull request closed without merging**